src/optimize.ml

```diff
@@ -0,0 +1,293 @@
+  match expr with
+  | ((I32_const _ | I64_const _) as x)
+    :: ((I32_const _ | I64_const _) as y)
+    :: (I_binop (nn, op) as i_binop)
+    :: tl -> begin
+    try
+      let result =
+        Interpret.exec_ibinop [ Value.of_instr y; Value.of_instr x ] nn op
+      in
+      begin
+        match result with
+        | [ ((I32 _ | I64 _) as result) ] ->
+          optimize_expr (Value.to_instr result :: tl)
+        | _ -> assert false
+      end
+    with Trap _ -> x :: optimize_expr (y :: i_binop :: tl)
+  end
+  | ((F32_const _ | F64_const _) as x)
+    :: ((F32_const _ | F64_const _) as y)
+    :: F_binop (nn, op)
+    :: tl ->
+    let result =
+      Interpret.exec_fbinop [ Value.of_instr y; Value.of_instr x ] nn op
+    in
+    begin
+      match result with
+      | [ ((F32 _ | F64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | ((I32_const _ | I64_const _) as x) :: I_unop (nn, op) :: tl ->
+    let result = Interpret.exec_iunop [ Value.of_instr x ] nn op in
+    begin
+      match result with
+      | [ ((I32 _ | I64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | ((F32_const _ | F64_const _) as x) :: F_unop (nn, op) :: tl ->
+    let result = Interpret.exec_funop [ Value.of_instr x ] nn op in
+    begin
+      match result with
+      | [ ((F32 _ | F64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | ((I32_const _ | I64_const _) as x) :: I_testop (nn, op) :: tl ->
+    let result = Interpret.exec_itestop [ Value.of_instr x ] nn op in
+    begin
+      match result with
+      | [ ((I32 _ | I64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | ((I32_const _ | I64_const _) as x)
+    :: ((I32_const _ | I64_const _) as y)
+    :: I_relop (nn, op)
+    :: tl ->
+    let result =
+      Interpret.exec_irelop [ Value.of_instr y; Value.of_instr x ] nn op
+    in
+    begin
+      match result with
+      | [ ((I32 _ | I64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | ((F32_const _ | F64_const _) as x)
+    :: ((F32_const _ | F64_const _) as y)
+    :: F_relop (nn, op)
+    :: tl ->
+    let result =
+      Interpret.exec_frelop [ Value.of_instr y; Value.of_instr x ] nn op
+    in
+    begin
+      match result with
+      | [ ((F32 _ | F64 _) as result) ] ->
+        optimize_expr (Value.to_instr result :: tl)
+      | _ -> assert false
+    end
+  | (I32_const _ | I64_const _ | F32_const _ | F64_const _) :: Drop :: tl ->
+    optimize_expr tl
+  | Local_set x :: Local_get y :: tl when x = y ->
+    optimize_expr (Local_tee x :: tl)
+  | (Br _ as br) :: _tl -> [ br ]
+  | (I32_const c as const) :: (Br_if _ as brif) :: tl -> begin
+    match c with 0l -> optimize_expr tl | _ -> [ const; brif ]
```

Comment:
You can replace `[ const; brif ]` by `[ Br label ]` (you'll need to name the label when matching on `Br_if`).